website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+It will:
+
+
+
+1. Authenticate - ensure you are allowed to perform actions on the Flink cluster
+2. Validate (Fetches template from AWS/GCS for validation) - will take user variable input and validate it against the policy and rules and yaml format.
+3. Execute the action - convert the java call to invoke kubernetes operation
+
+Lessons learned:
+
+
+
+1. App specific operator service - at our large scale, the operator was not able to handle such a large amount of jobs. Kubernetes calls started to time out and failed. To solve this, we created multiple operators (~4) in high traffic regions to handle each application.
+2. Kube call caching - Cache results of kubernetes calls for 30-60s to prevent overloading
+3. Label support - providing label support to search jobs via client specific variables reduced the load on Kube along with 5x faster search of the jobs
+
+Some of our biggest wins by exposing the library are:
+
+
+
+1. Standardized Job Management - Users can easily start, delete, and get status updates of their Flink jobs in a Kubernetes environment using a single library.
+2. Abstracted Kubernetes Complexity - Teams no longer need to worry about the inner workings of Kubernetes or formatting job deployment YAML files, as the library handles these details internally.
+3. Easy Upgrades - With the underlying Kubernetes infrastructure, the library brings robustness and fault tolerance to Flink job management, ensuring minimal downtime and efficient recovery.
+
+
+# Observability and Alerting
+
+Observability is important to run the production system at a large scale. We have ~30k streaming jobs in PANW, each one serves a customer for a specific application. Each job reads data from multiple topics in kafka, performs transformation and then writes the data to various sinks/endpoints.
+
+The constraint can be anywhere in the pipeline or its endpoints (Customer api, BigQuery, etc). We want to make sure the latency of streaming meets the SLA. Therefore, understanding if a job is healthy, meeting SLA or not, alerting/intervening if needed is very challenging.
+
+We build a sophisticated observability and alerting capability to achieve our operational goals. We provide 3 kinds of observability and debugging tools, described below.
+
+
+## Flink job list and job insight from Prometheus and Grafana
+
+Each Flink job sends various metrics to our Prometheus with great cardinality details, such as application name, customer id, regions, so that we can slice and dice to look at each job. Critical metrics include input traffic rate, output throughput, backlogs in Kafka, timestamp based latency, task CPU usage, task numbers, OOM counts, etc. Following charts are just a few examples: the charts give us details of ingestion traffic rate to Kafka for a specific customer, streaming job’s overall throughput and each vCPU’s throughput, backlogs in Kafka, worker autoscaling based on the observed backlog.
```

Comment:
```suggestion
Each Flink job sends various metrics to our Prometheus with cardinality details, such as application name, customer Id, and regions, so that we can look at each job. Critical metrics include the input traffic rate, output throughput, backlogs in Kafka, timestamp-based latency, task CPU usage, task numbers, OOM counts, and so on.

The following charts provide a few examples. The charts provide details about the ingestion traffic rate to Kafka for a specific customer, the streaming job’s overall throughput, each vCPU’s throughput, backlogs in Kafka, and worker autoscaling based on the observed backlog.
```